### PR TITLE
Increase codecov of pkg/kafka

### PR DESCRIPTION
**Pull request by hellspawn679**
## Which problem is this PR solving?
- https://github.com/jaegertracing/jaeger/issues/5068

## Description of the changes
- added test case for ` pkg/kafka/auth/config.go` and updated `internal/tracegen/worker_test.go`

## How was this change tested?
- 

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: pkg/kafka/auth/config_test.go [hunk 1]
```diff
@@ -1,82 +1,110 @@
...
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextUsername, "user")
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextPassword, "password")
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextMechanism, "SCRAM-SHA-256")
 
 	authConfig := &AuthenticationConfig{}
 	err := authConfig.InitFromViper(configPrefix, v)
 	require.NoError(t, err)
 	expectedConfig := &AuthenticationConfig{
-		Authentication: "kerberos",
+		Authentication: "tls",
```
Comment (yurishkuro) [inline pkg/kafka/auth/config_test.go:38]: what is the purpose of this change?

File: pkg/kafka/auth/config_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosRealm, "EXAMPLE.COM")
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosUseKeyTab, true)
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosUsername, "user")
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosPassword, "password")
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosConfig, "/path/to/krb5.conf")
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosKeyTab, "/path/to/keytab")
+	v.Set(configPrefix+kerberosPrefix+suffixKerberosDisablePAFXFAST, true)
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextUsername, "user")
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextPassword, "password")
+	v.Set(configPrefix+plainTextPrefix+suffixPlainTextMechanism, "SCRAM-SHA-256")
```
Comment (yurishkuro) [inline pkg/kafka/auth/config_test.go:32]: We don't test flags like this. Flags should be passed as literal strings, without using variables, as an array of CLI arguments.

File: internal/tracegen/worker_test.go [hunk 3]
```diff
@@ -16,30 +16,63 @@ import (
...
+			},
+		}
+		expectedOutput := `{"level":"info","msg":"Worker 7 generated 7 traces"}` + "\n"
+
+		worker.simulateTraces()
+		assert.Equal(t, expectedOutput, buf.String())
+	})
+
+	// Second test case with Pause = 3 * time.Second
+	t.Run("Pause 3 * time.Second", func(t *testing.T) {
```
Comment (yurishkuro) [inline internal/tracegen/worker_test.go:50]: What does this add over the previous sub-test? Pause is a parameter, how does increasing the parameter change the coverage?

Comment (hellspawn679) [reply internal/tracegen/worker_test.go:50]: i was trying to test `if w.Pause != 0 ` condition

Comment (yurishkuro) [reply internal/tracegen/worker_test.go:50]: Original L35 already had `Pause: time.Second`, which is ` != 0`. How does adding `p=3` improve on that?

File: pkg/kafka/auth/config_test.txt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+
+
+func TestSetConfiguration(t *testing.T) {
+	logger, _ := zap.NewDevelopment()
+	saramaConfig := sarama.NewConfig()
+	v := viper.New()
+	configPrefix := "kafka.auth."
+	v.Set(configPrefix+suffixAuthentication, "kerberos")
```
Comment (yurishkuro) [inline pkg/kafka/auth/config_test.txt:25]: we don't test flags like that. See cmd/ingester/app/flags_test.go how we test flags.

File: internal/tracegen/worker_test.go (hunk 3 above)
Comment (hellspawn679) [reply internal/tracegen/worker_test.go:50]: image

File: pkg/kafka/auth/config_test.txt (hunk 4 above)
Comment (hellspawn679) [reply pkg/kafka/auth/config_test.txt:25]: yes i have currently changed it but I am having some goleaks `TestSetConfiguration`

File: internal/tracegen/worker_test.go [hunk 5]
```diff
@@ -16,30 +16,63 @@ import (
...
-			Service:  "stdout",
-			Debug:    true,
-			Firehose: true,
-		},
-	}
-	expectedOutput := `{"level":"info","msg":"Worker 7 generated 7 traces"}` + "\n"
-
-	worker.simulateTraces()
-	assert.Equal(t, expectedOutput, buf.String())
+	// First test case with Pause = time.Second
```
Comment (yurishkuro) [inline internal/tracegen/worker_test.go:19]: if the value of Pause is the only difference between two tests, it should be written as a table test instead of duplicating so much code.

File: plugin/storage/integration/kafka_test.go [hunk 6]
```diff
@@ -91,6 +93,10 @@ func (s *KafkaIntegrationTestSuite) initialize(t *testing.T) {
 	}
 	options.InitFromViper(v)
 	traceStore := memory.NewStore()
+	_, err = builder.CreateConsumer(logger, metrics.NullFactory, traceStore, options)
+	require.Error(t, err, "invalid version `0`")
```
Comment (yurishkuro) [inline plugin/storage/integration/kafka_test.go:97]: `"invalid version `0`"` this string is non-functional. If you want to test that the error has this text then use ErrorEquals or ErrorContains

File: plugin/storage/integration/kafka_test.go [hunk 7]
```diff
@@ -91,6 +93,10 @@ func (s *KafkaIntegrationTestSuite) initialize(t *testing.T) {
...
+	command.ParseFlags([]string{"--kafka.consumer.protocol-version", "2.0.0"})
```
Comment (yurishkuro) [inline plugin/storage/integration/kafka_test.go:98]: add new line to separate logically different group
```suggestion

	command.ParseFlags([]string{"--kafka.consumer.protocol-version", "2.0.0"})
```

File: pkg/kafka/auth/tls.go [hunk 8]
```diff
@@ -26,6 +26,8 @@ import (
 func setTLSConfiguration(config *tlscfg.Options, saramaConfig *sarama.Config, logger *zap.Logger) error {
 	if config.Enabled {
 		tlsConfig, err := config.Config(logger)
+		defer config.Close()
```
Comment (yurishkuro) [inline pkg/kafka/auth/tls.go:29]: what does `defer` do?

Comment (hellspawn679) [reply pkg/kafka/auth/tls.go:29]: it close `certWatcher` which uses `fswatcher` which was giving goleak issue that I talked about

Comment (yurishkuro) [reply pkg/kafka/auth/tls.go:29]: correct, but _when_ does it do that?

Comment (hellspawn679) [reply pkg/kafka/auth/tls.go:29]: after the test ends when tlscfg.Options is ended

Comment (yurishkuro) [reply pkg/kafka/auth/tls.go:29]: no. Read up on how defer works.

File: pkg/kafka/consumer/config_test.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	// v.BindPFlags(command.Flags())
+
+	// authConfig := &auth.AuthenticationConfig{}
+	// command.ParseFlags([]string{
+	// 	"--kafka.auth.authentication=fail",
+	// })
+
+	// err := authConfig.InitFromViper(configPrefix, v)
+	// require.NoError(t, err)
+	// require.Error(t, authConfig.SetConfiguration(saramaConfig, logger), "Unknown/Unsupported authentication method fail to kafka cluster")
```
Comment (hellspawn679) [inline pkg/kafka/consumer/config_test.go:43]: @yurishkuro Should I test this using a flag, despite the increase in code size, or should I opt for a shorter implementation?

File: internal/tracegen/worker_test.go [hunk 10]
```diff
@@ -16,30 +16,53 @@ import (
...
+		name           string
+		pause          time.Duration
+		childSpans     int
+		expectedOutput string
+	}{
+		{
+			name:           "no pause",
+			pause:          0,
+			childSpans:     1,
+			expectedOutput: `{"level":"info","msg":"Worker 7 generated 7 traces"}` + "\n",
```
Comment (yurishkuro) [inline internal/tracegen/worker_test.go:29]: if these are the same for all tests, no need to include them as params, it only increases cognitive overhead for understanding the test

File: pkg/kafka/auth/config_test.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+	"github.com/jaegertracing/jaeger/pkg/config/tlscfg"
+)
+
+func Test_InitFromViper(t *testing.T) {
+	configPrefix := "kafka.auth"
+	flagSet := flag.NewFlagSet("flags", flag.ContinueOnError)
+	AddFlags(configPrefix, flagSet)
+	command := &cobra.Command{}
+	command.Flags().AddGoFlagSet(flagSet)
```
Comment (yurishkuro) [inline pkg/kafka/auth/config_test.go:26]: we have pkg/config/Viperize for this

File: pkg/kafka/auth/config_test.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+		command.ParseFlags([]string{
+			"--kafka.auth.authentication=" + authType,
+			"--kafka.auth.kerberos.service-name=kafka",
+			"--kafka.auth.kerberos.realm=EXAMPLE.COM",
+			"--kafka.auth.kerberos.use-keytab=true",
+			"--kafka.auth.kerberos.username=user",
+			"--kafka.auth.kerberos.password=password",
+			"--kafka.auth.kerberos.config-file=/path/to/krb5.conf",
+			"--kafka.auth.kerberos.keytab-file=/path/to/keytab",
+			"--kafka.auth.kerberos.disable-fast-negotiation=true",
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply pkg/kafka/auth/config_test.go:105]: If you are testing different permutations of settings, it's better to make this table-driven and include only the right amount of settings for each use case

File: pkg/kafka/auth/config_test.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+			"--kafka.auth.kerberos.use-keytab=false",
+		})
+		authConfig = &AuthenticationConfig{}
+		err := authConfig.InitFromViper(configPrefix, v)
+		require.NoError(t, err)
+	}
+
+	// Test with invalid authentication method
+	parseFlagsAndInit("fail")
+	require.Error(t, authConfig.SetConfiguration(saramaConfig, logger), "Unknown/Unsupported authentication method fail to kafka cluster")
```
[the remark this answers is not in the document]
Comment (hellspawn679) [reply pkg/kafka/auth/config_test.go:119]: L73 of `pkg/kafka/auth/config.go`

File: pkg/kafka/auth/config_test.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+	parseFlagsAndInit(" ")
+	require.NoError(t, authConfig.SetConfiguration(saramaConfig, logger))
+
+	// Test with tls
+	parseFlagsAndInit("tls")
+	require.NoError(t, authConfig.SetConfiguration(saramaConfig, logger))
+	defer authConfig.TLS.Close()
+	// test tls_fail
+	authConfig.TLS.CipherSuites = []string{"fail"}
+	require.Error(t, authConfig.SetConfiguration(saramaConfig, logger))
```
[the remark this answers is not in the document]
Comment (hellspawn679) [reply pkg/kafka/auth/config_test.go:141]: L30 in ` pkg/kafka/auth/tls.go`

File: pkg/kafka/auth/config_test.go [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+	// Test with kerberos
+	parseFlagsAndInit("kerberos")
+	require.NoError(t, authConfig.SetConfiguration(saramaConfig, logger))
+
+	// Test all plaintext options
+	parseFlagsAndInit("plaintext")
+	testPlaintext(v, t, configPrefix, logger, "SCRAM-SHA-256", saramaConfig)
+	testPlaintext(v, t, configPrefix, logger, "SCRAM-SHA-512", saramaConfig)
+	testPlaintext(v, t, configPrefix, logger, "PLAIN", saramaConfig)
```
[the remark this answers is not in the document]
Comment (hellspawn679) [reply pkg/kafka/auth/config_test.go:129]: it is handled by the func  `testPlaintext`

File: pkg/kafka/auth/tls_test.go [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/jaegertracing/jaeger/pkg/config/tlscfg"
+)
+
+func TestSetTLSConfiguration(t *testing.T) {
+	logger, _ := zap.NewDevelopment()
+	saramaConfig := sarama.NewConfig()
+	tlsConfig := &tlscfg.Options{
+		Enabled: true,
+	}
+	defer tlsConfig.Close()
```
Comment (yurishkuro) [inline pkg/kafka/auth/tls_test.go:23]: this should be after the call to `setTLSConfiguration` and error check. At this point you just created a struct, there can be no goroutine leak yet.

File: pkg/kafka/auth/tls_test.go [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/Shopify/sarama"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/pkg/config/tlscfg"
+)
+
+func TestSetTLSConfiguration(t *testing.T) {
+	logger, _ := zap.NewDevelopment()
```
Comment (yurishkuro) [inline pkg/kafka/auth/tls_test.go:18]:
```suggestion
	logger := zaptest.NewLogger(t)
```
please use it everywhere in new tests

File: plugin/storage/integration/kafka_test.go [hunk 18]
```diff
@@ -91,6 +100,11 @@ func (s *KafkaIntegrationTestSuite) initialize(t *testing.T) {
 	}
 	options.InitFromViper(v)
 	traceStore := memory.NewStore()
+	_, err = builder.CreateConsumer(logger, metrics.NullFactory, traceStore, options)
+	require.EqualError(t, err, "invalid version `0`")
```
Comment (yurishkuro) [inline plugin/storage/integration/kafka_test.go:104]: you already tested this condition in L68, what's the value of doing it again?

File: pkg/kafka/auth/plaintext_test.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+func TestSetPlainTextConfiguration(t *testing.T) {
+	tests := []struct {
+		name              string
+		config            PlainTextConfig
+		expectedError     error
+		expectedMechanism sarama.SASLMechanism
+	}{
+		{
+			name: "SCRAM-SHA-256",
```
Comment (yurishkuro) [inline pkg/kafka/auth/plaintext_test.go:42]: `name` field is redundant, you can use `config.Mechanism` instead

**Review by yurishkuro — APPROVED** [review]
